src/dataframe/insertion.jl

```diff
@@ -0,0 +1,958 @@
+
+julia> insert!(df, 3, NamedTuple(), cols=:subset)
+8×3 DataFrame
+ Row │ A        B        C
+     │ Any      Int64?   Float64?
+─────┼─────────────────────────────
+   1 │ 11            12  missing
+   2 │ a              1  missing
+   3 │ missing  missing  missing
+   4 │ true           0  missing
+   5 │ b              2  missing
+   6 │ c              3  missing
+   7 │ a              1  missing
+   8 │ 1.0      missing        1.0
+```
+"""
+insert!
+
+function Base.insert!(df::DataFrame, loc::Integer, row::Any; promote::Bool=false)
+    loc isa Bool && throw(ArgumentError("invalid index: $loc of type Bool"))
+    1 <= loc <= nrow(df)+1 ||
+        throw(ArgumentError("invalid index: $loc for data frame with $(nrow(df)) rows"))
+    return _row_inserter!(df, loc, row, Val{:insert}(), promote)
+end
+
+function _row_inserter!(df::DataFrame, loc::Integer, row::Any,
+                        mode::Union{Val{:push}, Val{:pushfirst}, Val{:insert}},
+                        promote::Bool)
+    if !(row isa Union{Tuple, AbstractArray})
+        # an explicit error is thrown as this was allowed in the past
+        throw(ArgumentError("it is not allowed to insert collections of type " *
+                            "$(typeof(row)) into a DataFrame. Only " *
+                            "`Tuple`, `AbstractArray`, `AbstractDict`, `DataFrameRow` " *
+                            "and `NamedTuple` are allowed."))
+    end
+    nrows, ncols = size(df)
+    targetrows = nrows + 1
+    if length(row) != ncols
+        msg = "Length of `row` does not match `DataFrame` column count."
+        throw(DimensionMismatch(msg))
+    end
+    current_col = 0
+    try
+        for (i, (col, val)) in enumerate(zip(_columns(df), row))
+            current_col += 1
+            @assert length(col) == nrows
+            S = typeof(val)
+            T = eltype(col)
+            if S <: T || !promote || promote_type(S, T) <: T
+                mode isa Val{:push} && push!(col, val)
+                mode isa Val{:pushfirst} && pushfirst!(col, val)
+                mode isa Val{:insert} && insert!(col, loc, val)
+            else
+                newcol = Tables.allocatecolumn(promote_type(S, T), targetrows)
+                firstindex(newcol) != 1 && _onebased_check_error()
+                if mode isa Val{:push}
+                    copyto!(newcol, 1, col, 1, nrows)
+                    newcol[end] = val
+                end
+                if mode isa Val{:pushfirst}
+                    newcol[1] = val
+                    copyto!(newcol, 2, col, 1, nrows)
+                end
+                if mode isa Val{:insert}
+                    copyto!(newcol, 1, col, 1, loc-1)
+                    newcol[loc] = val
+                    copyto!(newcol, loc+1, col, loc, nrows-loc+1)
+                end
+                _columns(df)[i] = newcol
+            end
+        end
+    catch err
+        #clean up partial row
+        for col2 in _columns(df)[1:current_col]
```

Comment:
Let's avoid a small allocation? Same elsewhere.
```suggestion
        for col2 in view(_columns(df), 1:current_col)
```
or
```suggestion
        for i in 1:current_col
            col2 = _columns(df)[i]
```

Reply:
OK - I guess it is good to change for the cleanness sake. This code is only invoked on error, so we do not care if it is fast.

Reply:
Ah right. Maybe not worth changing then.

Reply:
I have changed this